WT-5212 incremental backup data validation tests

raviprakashgiri29 opened the pull request.

sueloverso commented:
@raviprakashgiri29 the general ideas here seem fine. However, I don't think the example program is the place for testing. I pointed you to it as a place that showed how to use the new interface. Ideally, a test with specific settings like bulk load would be a Python test. There are currently 2 incremental related python tests:
* test_backup11.py tests the API, including using the API in error situations.
* test_backup12.py is a test that uses the API but does not do the block copies. It opens the incremental cursor and gets the file name and checks the type, but always does a full file copy.

I suggest either adding one or more new python tests or adding a new C test program in `test/csuite`. The example program should be kept as clean as possible, and it is already pretty complex.

raviprakashgiri29 commented:
> @raviprakashgiri29 the general ideas here seem fine. However, I don't think the example program is the place for testing. I pointed you to it as a place that showed how to use the new interface. Ideally, a test with specific settings like bulk load would be a Python test. There are currently 2 incremental related python tests:
> 
> * test_backup11.py tests the API, including using the API in error situations.
> * test_backup12.py is a test that uses the API but does not do the block copies. It opens the incremental cursor and gets the file name and checks the type, but always does a full file copy.
> 
> I suggest either adding one or more new python tests or adding a new C test program in `test/csuite`. The example program should be kept as clean as possible, and it is already pretty complex.

@sueloverso Thank you for the suggestion. I will try to include all the scenarios in a new C test program under 'test/csuite' and leave `ex_backup_block.c` as it is.

sueloverso commented on test/csuite/wt5212_incremental_backup/main.c:
```diff
@@ -0,0 +1,779 @@
+    const char *filename;
+    bool first;
+
+    /*! [incremental backup using block transfer]*/
```
This is a documentation label from the example. It should be removed.

raviprakashgiri29 replied:
Fixed

sueloverso commented on test/csuite/wt5212_incremental_backup/main.c:
```diff
@@ -0,0 +1,779 @@
+    testutil_check(backup_cur->close(backup_cur));
+    finalize_files(flist, count);
+    free(tmp);
+    /*! [incremental backup using block transfer]*/
```
Remove this too.

raviprakashgiri29 replied:
Fixed

sueloverso commented on test/csuite/wt5212_incremental_backup/main.c:
```diff
@@ -0,0 +1,779 @@
+    /*! [incremental backup using block transfer]*/
+}
+
+static void
```
This function should have a comment.

raviprakashgiri29 replied:
Fixed

sueloverso commented on test/csuite/wt5212_incremental_backup/main.c:
```diff
@@ -0,0 +1,779 @@
+}
+
+/*
+ * Remove all the records, take backups and validate
```
All comments should be complete sentences, ending in a period.

raviprakashgiri29 replied:
Fixed

sueloverso commented on test/csuite/wt5212_incremental_backup/main.c:
```diff
@@ -0,0 +1,779 @@
+    testutil_check(compare_backups(i, uri));
+}
+
+static void
```
Again add a function comment.

raviprakashgiri29 replied:
Fixed

sueloverso commented on test/csuite/wt5212_incremental_backup/main.c:
```diff
@@ -0,0 +1,779 @@
+}
+
+static void
+drop_old_add_new_objects(WT_CONNECTION *wt_conn)
```
To me it is clearer to say `table` instead of `object`.

raviprakashgiri29 replied:
Fixed

sueloverso commented on test/csuite/wt5212_incremental_backup/main.c:
```diff
@@ -0,0 +1,779 @@
+
+    new_object = true;
+
+    for (i = ITERATIONS_MULTIPLIER + 1; i < ITERATIONS_MULTIPLIER * 2; i++) {
```
The loop bounds are not intuitive and confusing. They feel like they're fragile, but I really don't know what you're expressing here. So either it should be clearly commented or think about whether it could be simplified. What does the loop give us? It seems like you could just do the drop and create and then add data to uri3 once and that would be sufficient.

raviprakashgiri29 replied:
I have updated the code to remove irrelevant loops and added a global variable `"counter"` to drive this.
Note - The max value of `counter` would be MAX_ITERATIONS.

sueloverso commented on test/csuite/wt5212_incremental_backup/main.c:
```diff
@@ -0,0 +1,779 @@
+}
+
+/*
+ * Create previously dropped object and add new contents to it
```
Again, end the sentence with punctuation and `object` -> `table`.

raviprakashgiri29 replied:
Fixed

sueloverso commented on test/csuite/wt5212_incremental_backup/main.c:
```diff
@@ -0,0 +1,779 @@
+    testutil_check(wt_conn->open_session(wt_conn, NULL, NULL, &session));
+    testutil_check(session->create(session, uri2, "key_format=S,value_format=S"));
+
+    for (i = ITERATIONS_MULTIPLIER * 2; i < ITERATIONS_MULTIPLIER * 3; i++) {
```
Again, I would think about if the loop is necessary or if one time through would be enough for testing.

raviprakashgiri29 replied:
I have changed the code to run once. Thanks.

sueloverso commented on test/csuite/wt5212_incremental_backup/main.c:
```diff
@@ -0,0 +1,779 @@
+    add_bulk_load(session, 0, 0, uri4);
+    // ID 15 is used for logged table
+    take_full_backup(session, ITERATIONS_MULTIPLIER * 3);
+    take_incr_backup(session, ITERATIONS_MULTIPLIER * 3);
```
This feels fragile and could break in unexpected ways if someone adds other things.

raviprakashgiri29 replied:
Fixed.

sueloverso commented on test/csuite/wt5212_incremental_backup/main.c:
```diff
@@ -0,0 +1,779 @@
+    take_full_backup(session, 16);
+    take_incr_backup(session, 16);
+    printf("Dumping and comparing data\n");
+    testutil_check(compare_backups(16, uri5));
```
This feels even more fragile.

raviprakashgiri29 replied:
Fixed.

sueloverso commented on test/csuite/wt5212_incremental_backup/main.c:
```diff
@@ -0,0 +1,779 @@
+    testutil_check(session->open_cursor(session, "backup:", NULL, cmd_buf, &backup_cur));
+    testutil_check(backup_cur->close(backup_cur));
+
+    /*
```
Everything that's doing backup from here to the end should not be needed.

raviprakashgiri29 replied:
Fixed

sueloverso commented on test/csuite/wt5212_incremental_backup/main.c:
```diff
@@ -0,0 +1,779 @@
+    testutil_check(session->create(session, uri3, "key_format=S,value_format=S"));
+
+    testutil_check(session->open_cursor(session, "backup:", NULL, NULL, &backup_cur));
+    testutil_check(backup_cur->close(backup_cur));
```
What is the purpose of opening and then immediately closing this cursor here?

raviprakashgiri29 replied:
I agree it is a redundant code. Removed.

sueloverso commented on test/csuite/wt5212_incremental_backup/main.c:
```diff
@@ -0,0 +1,779 @@
+
+    testutil_check(wt_conn->close(wt_conn, NULL));
+    testutil_check(wiredtiger_open(home, NULL, CONN_CONFIG, &wt_conn));
+    testutil_check(wt_conn->open_session(wt_conn, NULL, NULL, &session));
```
Why does this function, and the next function, both close and reopen the connection and session? Does it need to do that? Does that add something to the test?

raviprakashgiri29 replied:
It throws `"Device or resource busy"` error if I try to drop an object without doing this. So, I have updated the code to do this just before dropping a table i.e, before `drop_old_add_new_table` function and removed from other places.

sueloverso commented on test/csuite/wt5212_incremental_backup/main.c:
```diff
@@ -0,0 +1,779 @@
+    WT_CURSOR *cursor;
+    int i;
+    int k;
+    char v[64];
```
This is repeat code from `add_work` mostly. Can this be refactored so that this function and `add_work` both call some other function like `add_data` and take cursors as an arg?

raviprakashgiri29 replied:
I have updated the `add_work` function to take an extra boolean argument "bulk_load" to drive the bulk load option.

sueloverso commented on test/csuite/wt5212_incremental_backup/main.c:
```diff
@@ -0,0 +1,779 @@
+        testutil_check(__wt_snprintf(k, sizeof(k), "key.%d.%d.%d", iter_i, iter_j, i));
+        testutil_check(__wt_snprintf(v, sizeof(v), "value.%d.%d.%d", iter_i, iter_j, i));
+        cursor->set_key(cursor, k);
+        cursor->set_value(cursor, v);
```
The `remove` call only needs a key, not a value.

raviprakashgiri29 replied:
Fixed

sueloverso commented:
@raviprakashgiri29 I have fixed `test_backup12.py` to correctly use the incremental backup API (as in copy the ranges when given). I think it might be easier than the C code for some of the individual things you want to test. It is easier to maintain the file lists and things like that. You can find the new test in PR https://github.com/wiredtiger/wiredtiger/pull/5280.

raviprakashgiri29 commented on test/csuite/wt5212_incremental_backup/main.c:
```diff
@@ -0,0 +1,779 @@
+#define CONN_CONFIG "create,cache_size=100MB,log=(enabled=true,path=logpath,file_max=100K)"
+// Max iterations are (ITERATIONS_MULTIPLIER * 3) + 2 = 17
+#define ITERATIONS_MULTIPLIER 5
+#define MAX_ITERATIONS 17
```
Fixed.

raviprakashgiri29 commented:
> This has all the parts, but I think it needs some changes. This only added more to the example program. I think you should really review what that does and whether it is necessary here. Some thoughts:
> 
> * The example program has a lot of the basic functions you need, such as taking the full or incremental backup and how to compare content and manage the file list.
> * However it also has a lot of extra stuff you probably don't want here. The N iterations that the example does is probably not necessary. Maybe 1 iteration, maybe don't do it at all.
> * I suggest reconsidering the directory management. Instead of numbering them for loops, give each subtest a name like "DROP", "RECREATE", "BULK", and that will help the fragile comment I made.
> * Simplify and have this test do only what it needs to. Please also add a comment describing the cases it is testing.

Thank you @sueloverso for reviewing the changes. I have updated this file to remove redundant code and put comments wherever possible.

raviprakashgiri29 commented:
> @raviprakashgiri29 I have fixed `test_backup12.py` to correctly use the incremental backup API (as in copy the ranges when given). I think it might be easier than the C code for some of the individual things you want to test. It is easier to maintain the file lists and things like that. You can find the new test in PR #5280.

@sueloverso Thank you for updating `test_backup12.py`. I reviewed the file and it does the incremental block copy. I have updated the C test as suggested and addressed the review comments. Please review and let me know if I can go ahead with the C test or migrate into Python test file.

sueloverso commented on test/csuite/wt5212_incremental_backup/main.c:
```diff
@@ -0,0 +1,659 @@
+    if (bulk_load) {
+        testutil_check(session->open_cursor(session, t_uri, NULL, "bulk", &cursor));
+    } else {
+        testutil_check(session->open_cursor(session, t_uri, NULL, NULL, &cursor));
```
Instead of passing in a bool you could just pass in the string of NULL or "bulk" and then avoid the conditional.

sueloverso commented on test/csuite/wt5212_incremental_backup/main.c:
```diff
@@ -0,0 +1,659 @@
+    /*
+     * Close and re-open the connection to drop existing table.
+     */
+    testutil_check(wt_conn->close(wt_conn, NULL));
```
You still should not need to do this. I think the EBUSY error is likely to be a cursor left open on the table you want to drop. Check for a cursor being left open on this table somewhere. It is not harmful exactly, but should not be necessary.

sueloverso commented on test/csuite/wt5212_incremental_backup/main.c:
```diff
@@ -0,0 +1,659 @@
+remove_all_records_validate(WT_SESSION *session)
+{
+    remove_work(session);
+    take_full_backup(session);
```
I think the intent for all these cases is that we don't need to take the full backup, but we do the operation and take only an incremental.

raviprakashgiri29 commented:
> I have a few comments on the C program. It looks a lot better now and cleaner.
> 
> While I know you have done a lot of work on this program, I do think these scenarios would be better as small python tests. This program does what you've set out to do, but it feels like it will be harder in the future to add new cases to the C program. Adding new python scenarios or tests is often a lot easier.

@sueloverso Thank you for reviewing the changes. 

I will close this pull request, migrate the changes to the Python test and create a new pull request.
